### PR TITLE
Tbp/google rate limit exceeded

### PR DESCRIPTION
**Pull request by TheBestPessimist**
With this i want to fix/delay #123  and #124.

This is the implementation i have come up with regarding the backoff computation for GCD storage.

Maybe this could be made in  a standalone file, and applied somehow to all storages?

### CODE REVIEW

File: src/duplicacy_gcdstorage.go [hunk 1]
```diff
@@ -27,10 +27,11 @@ import (
...
 
-	service     *drive.Service
-	idCache     map[string]string
-	idCacheLock *sync.Mutex
-	backoffs    []int
+	service         *drive.Service
+	idCache         map[string]string
+	idCacheLock     *sync.Mutex
+	backoffs        []float64
+	backoffsRetries []int
```
Comment (tbain98) [inline src/duplicacy_gcdstorage.go]: Because the variable name could be interpreted ambiguously, I recommend you add a comment to say that this is the number of backoffs that have been performed so far for the current send attempt on the given thread. Also, it's a small thing, but I think the variable name would be more clear as backoffRetries (or as backoffRetriesPerThread) than as backoffsRetries.

File: src/duplicacy_gcdstorage.go [hunk 2]
```diff
@@ -45,11 +46,19 @@ type GCDConfig struct {
 }
 
 func (storage *GCDStorage) shouldRetry(threadIndex int, err error) (bool, error) {
+	const LIMIT_BACKOFF_TIME = 64
```
Comment (tbain98) [inline src/duplicacy_gcdstorage.go]: This whole 6-line block looks like it could be pulled out as member variables (or consts), because they don't appear to change between method calls. That would avoid spending a few CPU cycles on each method call, but also slim down the method body to focus on the actual algorithm.

File: src/duplicacy_gcdstorage.go [hunk 3]
```diff
@@ -62,8 +71,9 @@ func (storage *GCDStorage) shouldRetry(threadIndex int, err error) (bool, error)
 			retry = true
 		} else if e.Code == 403 {
 			// User Rate Limit Exceeded
-			message = "User Rate Limit Exceeded"
+			message = e.Message // "User Rate Limit Exceeded"
```
Comment (tbain98) [inline src/duplicacy_gcdstorage.go]: Looks like you accidentally left the previous message commented out; please remove.

File: src/duplicacy_gcdstorage.go [hunk 4]
```diff
@@ -62,8 +71,9 @@ func (storage *GCDStorage) shouldRetry(threadIndex int, err error) (bool, error)
...
 			retry = true
+
```
Comment (tbain98) [inline src/duplicacy_gcdstorage.go]: Superfluous and not consistent with the surrounding code; remove.

File: src/duplicacy_gcdstorage.go [hunk 5]
```diff
@@ -83,18 +93,37 @@ func (storage *GCDStorage) shouldRetry(threadIndex int, err error) (bool, error)
...
+		storage.backoffs[threadIndex] = computeInitialBackoff(minimumSleep, maximumSleep)
+		storage.backoffsRetries[threadIndex] = 0
 		return false, err
 	}
 
-	delay := float32(storage.backoffs[threadIndex]) * rand.Float32()
-	LOG_DEBUG("GCD_RETRY", "%s; retrying after %.2f seconds", message, delay)
-	time.Sleep(time.Duration(float32(storage.backoffs[threadIndex]) * float32(time.Second)))
-	storage.backoffs[threadIndex] *= 2
+	if storage.backoffs[threadIndex] < LIMIT_BACKOFF_TIME {
```
Comment (tbain98) [inline src/duplicacy_gcdstorage.go]: This logic is going to result in you going over your limit, which isn't the goal. You also have a bug where you're only incrementing the retries count when you're over the limit.

Instead, do the following:

```storage.backoffs[threadIndex] *= 2
if storage.backoffs[threadIndex] > LIMIT_BACKOFF_TIME {
    storage.backoffs[threadIndex] = LIMIT_BACKOFF_TIME
}
storage.backoffsRetries[threadIndex]++
```

File: src/duplicacy_gcdstorage.go [hunk 6]
```diff
@@ -83,18 +93,37 @@ func (storage *GCDStorage) shouldRetry(threadIndex int, err error) (bool, error)
...
+		storage.backoffs[threadIndex] *= 2.0
+	} else {
+		storage.backoffs[threadIndex] = LIMIT_BACKOFF_TIME
+		storage.backoffsRetries[threadIndex] += 1
+	}
+	delay := storage.backoffs[threadIndex]*rand.Float64() + storage.backoffs[threadIndex]*rand.Float64()
```
Comment (tbain98) [inline src/duplicacy_gcdstorage.go]: This merits an explanatory comment.

File: src/duplicacy_gcdstorage.go [hunk 7]
```diff
@@ -45,11 +46,19 @@ type GCDConfig struct {
 }
 
 func (storage *GCDStorage) shouldRetry(threadIndex int, err error) (bool, error) {
+	const LIMIT_BACKOFF_TIME = 64
+	const MAX_NUMBER_OF_RETRIES = 15
+	minimumSleepRatio := 0.1
+	maximumSleepRatio := 0.2
+	minimumSleep := float64(storage.numberOfThreads) * minimumSleepRatio
```
Comment (tbain98) [inline src/duplicacy_gcdstorage.go]: How about a comment explaining why you're using this algorithm (threads * 0.1 or 0.2) and what this particular algorithm is trying to accomplish. As-is, a casual reviewer has no way to know why we're doing this nor whether it's the right thing to be doing.

File: src/duplicacy_gcdstorage.go [hunk 8]
```diff
@@ -83,18 +93,37 @@ func (storage *GCDStorage) shouldRetry(threadIndex int, err error) (bool, error)
...
+		storage.backoffsRetries[threadIndex] += 1
+	}
+	delay := storage.backoffs[threadIndex]*rand.Float64() + storage.backoffs[threadIndex]*rand.Float64()
+	LOG_DEBUG("GCD_RETRY", "Thread: %3d. Message: %s. Retrying after %6.2f seconds. Current backoff: %6.2f. Number of retries: %2d.", threadIndex, message, delay, storage.backoffs[threadIndex], storage.backoffsRetries[threadIndex])
+	time.Sleep(time.Duration(delay * float64(time.Second)))
+
 	return true, nil
 }
 
+/*
```
Comment (tbain98) [inline src/duplicacy_gcdstorage.go]: Although I applaud the explanatory comments, I can't help feeling that most of this content is in the wrong place. In this method you've implemented an algorithm to choose a random number between min and max, but 1) there's nothing in this method about 0.1 and 0.2, that part of the comment belongs on the code that uses those values, 2) although the SO reference could be useful, you'd be better off just leaving it off and writing in plain English that you're picking a random number between min (inclusive) and max (inclusive), 3) it's not at all clear why Google won't be DDOS'ed because of the choice to sleep between min and max (or is this still part of the commentary on threads * 0.1 or 0.2, covered by my comment 1?).

File: src/duplicacy_gcdstorage.go [hunk 9]
```diff
@@ -274,7 +303,12 @@ func CreateGCDStorage(tokenFile string, storagePath string, threads int) (storag
 		numberOfThreads: threads,
 		idCache:         make(map[string]string),
 		idCacheLock:     &sync.Mutex{},
-		backoffs:        make([]int, threads),
+		backoffs:        make([]float64, threads),
+		backoffsRetries: make([]int, threads),
+	}
+
+	for b := range storage.backoffs {
+		storage.backoffs[b] = 0.1 * float64(storage.numberOfThreads) // at the first error, we should still sleep some amount
```
Comment (tbain98) [inline src/duplicacy_gcdstorage.go]: Why isn't this calling `computeInitialBackoff()`?

File: src/duplicacy_gcdstorage.go [hunk 10]
```diff
@@ -45,11 +46,19 @@ type GCDConfig struct {
 }
 
 func (storage *GCDStorage) shouldRetry(threadIndex int, err error) (bool, error) {
+	const LIMIT_BACKOFF_TIME = 64
+	const MAX_NUMBER_OF_RETRIES = 15
+	minimumSleepRatio := 0.1
+	maximumSleepRatio := 0.2
+	minimumSleep := float64(storage.numberOfThreads) * minimumSleepRatio
+	maximumSleep := float64(storage.numberOfThreads) * maximumSleepRatio
+	rand.Seed(time.Now().UnixNano()) // unsure if this is needed
```
Comment (tbain98) [inline src/duplicacy_gcdstorage.go]: It's needed once per program execution, but not once per method call, otherwise you're just wasting CPU cycled re-initializing it each time. It should be called near the top of `main()` if it's not already being called there, and then not called here.

**Comment by TheBestPessimist** [comment]
Will work on the changes in a few days. Right now I'm on a 15 hours plane ride. Bummer.

**Comment by gilbertchen** [comment]
I ran a few tests with this PR and it has been working great -- with 64 threads and a very small chunk size (1k) it kept going despite tons of 'rate exceeded' messages.  Thanks for the great work!

I'll make some minor changes after the merge, mostly in response to @tbain98's review comments.

**Comment by TheBestPessimist** [comment]
Ah, i forgot about this, thought it was already merged.  Sorry for not fixing everything. If i do it tonight, could you merge that as well, or it's easy to late?

**Comment by gilbertchen** [comment]
I created PR #236. Please review the changes I made there.
